server/src/test/java/org/elasticsearch/indices/state/ReopenWhileClosingIT.java

```diff
@@ -0,0 +1,167 @@
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import static java.util.Collections.emptySet;
+import static java.util.Collections.singletonList;
+import static org.elasticsearch.cluster.metadata.MetaDataIndexStateService.INDEX_CLOSED_BLOCK_ID;
+import static org.elasticsearch.indices.state.CloseIndexIT.assertIndexIsClosed;
+import static org.elasticsearch.indices.state.CloseIndexIT.assertIndexIsOpened;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.notNullValue;
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, minNumDataNodes = 2)
+public class ReopenWhileClosingIT extends ESIntegTestCase {
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return singletonList(MockTransportService.TestPlugin.class);
+    }
+
+    public void testReopenDuringClose() throws Exception {
+        final String indexName = "test";
+        createIndexWithDocs(indexName);
+
+        ensureClusterSizeConsistency();
+
+        final CountDownLatch block = new CountDownLatch(1);
+        final Releasable releaseBlock = interceptVerifyShardBeforeCloseActions(indexName, block::countDown);
+
+        ActionFuture<AcknowledgedResponse> closeIndexResponse = client().admin().indices().prepareClose(indexName).execute();
+        assertTrue("Waiting for index to have a closing blocked", block.await(60, TimeUnit.SECONDS));
+        assertIndexIsBlocked(indexName);
+        assertFalse(closeIndexResponse.isDone());
+
+        assertAcked(client().admin().indices().prepareOpen(indexName));
+
+        releaseBlock.close();
+        closeIndexResponse.get();
+        assertIndexIsOpened(indexName);
+    }
+
+    public void testReopenDuringCloseOnMultipleIndices() throws Exception {
+        final List<String> indices = new ArrayList<>();
+        for (int i = 0; i < randomIntBetween(2, 10); i++) {
+            indices.add("index-" + i);
+            createIndexWithDocs(indices.get(i));
+        }
+
+        ensureClusterSizeConsistency();
+
+        final CountDownLatch block = new CountDownLatch(1);
+        final Releasable releaseBlock = interceptVerifyShardBeforeCloseActions(randomFrom(indices), block::countDown);
+
+        ActionFuture<AcknowledgedResponse> closeIndexResponse = client().admin().indices().prepareClose("index-*").execute();
+        assertTrue("Waiting for index to have a closing blocked", block.await(60, TimeUnit.SECONDS));
+        assertFalse(closeIndexResponse.isDone());
+        indices.forEach(ReopenWhileClosingIT::assertIndexIsBlocked);
+
+        final List<String> reopenedIndices = randomSubsetOf(indices);
+        assertAcked(client().admin().indices().prepareOpen(reopenedIndices.toArray(Strings.EMPTY_ARRAY)));
+
+        releaseBlock.close();
+        closeIndexResponse.get();
```

Comment:
should this throw an exception / convey a failed closing?

Reply:
The close action should not be acknowledged so we can check that.

When we'll improve the Close Index API response we can decide if we want to throw an exception or report an error message. For now I think we should not change the API and just not ack the response.

Reply:
ok, can we assert that acked == false here then?

Reply:
It has been changed to `assertFalse(closeIndexResponse.get().isAcknowledged());`, I think we're good